csharp/src/Microsoft.ML.OnnxRuntime/NativeOnnxTensorMemory.cs

```diff
@@ -171,8 +171,11 @@ public NativeOnnxTensorMemory(OrtValue ortValue)
                         }
                     }
                 }
-                // Transfer ownership
-                _ortValue = new OrtValue(ortValue.Disown());
+                // Transfer ownership, but only do so if the original OrtValue is already owned.
+                if (ortValue.IsOwned)
+                    _ortValue = new OrtValue(ortValue.Disown());
+                else
+                    _ortValue = new OrtValue(ortValue.Handle, false);
```

Comment:
What is the use case for this? It was always supposed to be owned as this class disposes of it. It can not dispose of it if it does not own it.

Reply:
As an optimization, the error function callback uses a collection of OrtValues that directly reference the feeds and fetches received in the error function.   For this reason, we do not want to dispose of them for the memory is still being used by the model.  

See line 488 of  https://github.com/MyCaffe/onnxruntime/blob/WinTrainingCSharp/orttraining/orttraining/runner/orttraining_c_api.cc

Also see line 171 of 
https://github.com/MyCaffe/onnxruntime/blob/WinTrainingCSharp/csharp/src/Microsoft.ML.OnnxTraining/TrainingParameters.cs

And line 67 of
https://github.com/MyCaffe/onnxruntime/blob/WinTrainingCSharp/csharp/src/Microsoft.ML.OnnxTraining/OrtValueCollection.cs

Reply:
The inference user uses this class to ensure that the buffer is alive until it is disposed. Therefore, it is an invariant that must be held at all times. If you do not want the memory be disposed then keep them around until the memory no longer needed. The other options is to attach an non-owned buffer to the output OrtValue.

---
In reply to: [544527905](https://github.com/microsoft/onnxruntime/pull/6141#discussion_r544527905) [](ancestors = 544527905)

Reply:
"The other options is to attach an non-owned buffer to the output OrtValue."  This is what I will need to do - do you have a simple example of this?

Reply:
[https://github.com/microsoft/onnxruntime/blob/master/csharp/src/Microsoft.ML.OnnxRuntime/FixedBufferOnnxValue.cs#L101CreateFromMemory<T>](https://github.com/microsoft/onnxruntime/blob/master/csharp/src/Microsoft.ML.OnnxRuntime/FixedBufferOnnxValue.cs#L101CreateFromMemory<T>) does it. 
Note that, it pins the managed buffer and it needs to be disposed to UnPin it. So in a way there is an ownership here but it is not owned by the OrtValue.

---
In reply to: [544581578](https://github.com/microsoft/onnxruntime/pull/6141#discussion_r544581578) [](ancestors = 544581578)

Reply:
If ownership is transferred why wouldn't the Dispose implementation for the OrtValue instance handle that? In C++ a class can have a unique_ptr member to ensure disposal, but if someone transfers ownership via release() the unique_ptr dtor handles that. 

---
In reply to: [544560802](https://github.com/microsoft/onnxruntime/pull/6141#discussion_r544560802) [](ancestors = 544560802,544527905)

Reply:
@skottmckay ,
There are some cases where, for efficiency, you may not want the ownership transferred as is the case when passing OrtValues back in the error function callback on line 507 of https://github.com/MyCaffe/onnxruntime/blob/WinTrainingCSharp/orttraining/orttraining/runner/orttraining_c_api.cc